File: leetcode/dp/Climbing_Stairs/java/SolutionMemo.java
// https://leetcode.com/problems/climbing-stairs/description/
// 70. Climbing Stairs

import java.util.HashMap;
import java.util.Map;

// Recursive + Memoization | Top-Down | Dynamic Programming
// Time Complexity: O(n) because we are using a memoization array to store the results of the subproblems
// Space Complexity: O(n) because we are using a memoization array to store the results of the subproblems
class SolutionMemo {
    // memoization map to store the results of the subproblems
    Map<Integer, Integer> memo;
    public SolutionMemo() {
        this.memo = new HashMap<>();
        // adding base cases to the memoization map
        this.memo.put(1, 1);
        this.memo.put(2, 2);
    }
    public int climbStairs(int n) {
        if (this.memo.containsKey(n)) {
            return this.memo.get(n);
        } else {
            int res = climbStairs(n-2) + climbStairs(n-1);
            this.memo.put(n, res);
            return res;
        }
    }
    public static void main(String[] args) {
        int n = 45, expected = 1836311903;
        System.out.println("Solution Memo");
        SolutionMemo solmemo = new SolutionMemo();
        int actual = solmemo.climbStairs(n);
        System.out.println(actual);
        
        if (actual == expected) {
            System.out.println("Test Passed");
        } else {
            System.out.println("Test Failed: Expected " + expected + ", Actual " + actual);
        }
    }
}

File: leetcode/dp/Climbing_Stairs/python/SolutionMemo.py
# https://leetcode.com/problems/climbing-stairs/
# 70. Climbing Stairs

# Recursion + Memoization Solution | Top-Down | Dynamic Programming
# Time Complexity: O(n) because we are using a memoization table to store the results of the subproblems
# Space Complexity: O(n) because we are using a memoization table to store the results of the subproblems
class Solution:
    def climbStairs(self, n: int) -> int:
        memo = {1: 1, 2: 2}
        def findNoOfWays(i) -> int:
            if i in memo:
                return memo[i]
            else:
                memo[i] = findNoOfWays(i-2) + findNoOfWays(i-1)
                return memo[i]
        return findNoOfWays(n)

if __name__ == "__main__":
    n = 45
    expected = 1836311903
    solution = Solution()
    actual = solution.climbStairs(n)
    if actual == expected:
        print("Test Passed")
    else:
        print(f"Test Failed: Expected {expected}, Actual {actual}")